pkg/frontend/admin_openshiftcluster_delete_managedresource.go

```diff
@@ -0,0 +1,67 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/go-chi/chi/v5"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+)
+
+func (f *frontend) postAdminOpenShiftDeleteManagedResource(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	err := f._postAdminOpenShiftClusterDeleteManagedResource(ctx, r, log)
+	adminReply(log, w, nil, nil, err)
+}
+
+func (f *frontend) _postAdminOpenShiftClusterDeleteManagedResource(ctx context.Context, r *http.Request, log *logrus.Entry) error {
+	resType, resName, resGroupName := chi.URLParam(r, "resourceType"), chi.URLParam(r, "resourceName"), chi.URLParam(r, "resourceGroupName")
+	managedResourceID := r.URL.Query().Get("managedresourceid")
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", resType, resName, resGroupName)
+	case err != nil:
+		return err
+	}
+
+	if !strings.Contains(strings.ToLower(managedResourceID), strings.ToLower(doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID)) {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "The resource %s is not within the cluster's managed resource group %s.", managedResourceID, doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID)
+	}
+
+	subscriptionDoc, err := f.getSubscriptionDocument(ctx, doc.Key)
+	if err != nil {
+		return err
+	}
+
+	a, err := f.azureActionsFactory(log, f.env, doc.OpenShiftCluster, subscriptionDoc)
+	if err != nil {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", err.Error())
```

Comment:
This is a great question, and got me thinking, because I was also in the space of "you're right, this does seem weird".

My guess is Tony was thinking in HTTP best practice. In this case, we've got a dependent, external service that fails - what's the appropriate thing to return to the caller of _our_ API? Consensus differs, but generally people say 500/503 is a general standard if you don't have a cache to pull from.

I think our team has thought about this case a bit differently, because we're in `/admin`. Only technologically-literate people (although, I know _my_ status is up for debate) and automation will be calling this API, right? Why wrap errors in "niceties"? Well, I think the question turns to: "if we automate this, is it better for me to return 500 or 429 when I see a dependent service returns 429?"... and I think I fall on Tony's side for this. 500 seems more appropriate - I think errors should be handled as "close" as possible to the original client (retries, caching, etc).

Looking at the code block right below this I think adds some weight to this too. For a 404 case - we see Tony specifically call it out because callers of _our_ API, even automation, would care that the ResourceID they sent does not exist. I think this is good to merge, and might even be a good example for us to improve the rigidity of our `/admin` endpoints.

TL;DR - I'm +1, excellent work Tony!

Reply:
Now that you bring it up, I probably just saw this done in one of the many Geneva Actions already written and brought it in to this one without much thought.  I think it makes more sense to just return err and not wrap the error in `NewCloudError` since the adminReply() looks to do that by default.